### PR TITLE
update typescript version to 5

### LINKED ISSUE #7144: abitype@npm:0.7.1 [20ff3] → >=4.9.4 ✘
#### Expected behavior
➤ YN0002: │ web3-eth-abi@npm:4.2.2 doesn't provide typescript (p6fa46), requested by abitype
- Not seeing the above message when yarn install

#### Actual behavior
➤ YN0002: │ web3-eth-abi@npm:4.2.2 doesn't provide typescript (p6fa46), requested by abitype
- see the above message when yarn install

#### Steps to reproduce the behavior

1. `yarn install`
2. web3-eth-abi@npm:4.2.2 doesn't provide typescript (p6fa46), requested by abitype
3. `yarn explain peer-requirements p6fa46`
4.  `abitype@npm:0.7.1 [20ff3] → >=4.9.4 ✘`

#### Logs
```
➤ YN0000: ┌ Resolution step
➤ YN0002: │ web3-eth-abi@npm:4.2.2 doesn't provide typescript (p6fa46), requested by abitype
➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code

yarn explain peer-requirements p6fa46
➤ YN0000: web3-eth-abi@npm:4.2.2 doesn't provide typescript, breaking the following requirements:

➤ YN0000: abitype@npm:0.7.1 [20ff3] → >=4.9.4 ✘
```

I think the problem is caused by using typescript version 4.7.4 on `web3-eth-abi`.

#### Environment
Node: v16.19.1
typescript: v4.9.5
web3.js: v4.10.0

### PR DESCRIPTION
**Pull request by Muhammad-Altabba**
## Description

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](https://github.com/web3/web3.js/blob/4.x/.github/CONTRIBUTING.md).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run lint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have linked Issue(s) with this PR in "Linked Issues" menu.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -137,7 +137,7 @@
 		"typedoc-plugin-markdown": "^3.17.0",
 		"typedoc-plugin-mdn-links": "^2.0.0",
 		"typedoc-plugin-merge-modules": "^5.1.0",
-		"typescript": "^4.7.4",
```
Comment (jdevcs) [inline package.json:140]: will any of breaking changes in TS will affect users: 
https://github.com/microsoft/TypeScript/wiki/API-Breaking-Changes

https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#breaking-changes-and-deprecations

Comment (Muhammad-Altabba) [reply package.json:140]: As we are updating at devDependencies, there should be no issues with the braking changes except with our code that I needed to update at few places for the detected issues.

But for the users of our lib there should be, theoretically no issues at all. However, there could be some very edge case that I am not aware of. But according to my search and up to my knowledge, there should be no issues for the users.

File: packages/web3-utils/src/validation.ts [hunk 2]
```diff
@@ -134,10 +134,10 @@ export const compareBlockNumbers = (blockA: BlockNumberOrTag, blockB: BlockNumbe
 	) {
 		return 0;
 	}
-	if (blockA === 'earliest' && blockB > 0) {
```
Comment (avkos) [inline packages/web3-utils/src/validation.ts:137]: I do not clearly understand how it is related to the typescript version.

**Review by avkos — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (jdevcs) [reply package.json:140]: types created as output of build using TS in this PR are fully backward compatible?

File: packages/web3-utils/src/validation.ts (hunk 2 above)
Comment (Muhammad-Altabba) [reply packages/web3-utils/src/validation.ts:137]: The new TypeScript version detected an error in comparing not a necessarily number with a number.   And in this PR I optimized the condition.

File: package.json (hunk 1 above)
Comment (Muhammad-Altabba) [reply package.json:140]: According to my reading, the generated JavaScript is only affected by the targeted ECMA Script. And for the same ECMA Script, the transpiled code is compatible. I checked few of the generated JavaScript source code and things seems almost identical and 100% compatible.

**Review by avkos — APPROVED** [review]

**Review by jdevcs — COMMENTED** [review]
@Muhammad-Altabba before merging into 4.x, pls merge latest 4.x in to this branch and update configs as there is a new package as well : https://github.com/web3/web3.js/tree/4.x/packages/web3-account-abstraction thanks

**Review by danforbes — APPROVED** [review]

**Comment by jdevcs** [comment]
As discussed dev release from this PR needs to be tested with a plugin.

File: packages/web3-eth-accounts/tsconfig.types.tsbuildinfo [hunk 3]
```diff
@@ -0,0 +1 @@
...860.ts","./src/common/eips/4345.ts","./src/common/eips/4399.ts","./src/common/eips/4844.ts","./src/common/eips/4895.ts","./src/common/eips/5133.ts","./src/common/eips/index.ts","./src/common/hardforks/arrowGlacier.ts","./src/common/hardforks/berlin.ts","./src/common/hardforks/byzantium.ts","./src/common/hardforks/chainstart.ts","./src/common/hardforks/constantinople.ts","./src/common/hardforks/dao.ts","./src/common/hardforks/grayGlacier.ts","./src/common/hardforks/homestead.ts","./src/common/hardforks/index.ts","./src/common/hardforks/istanbul.ts","./src/common/hardforks/london.ts","./src/common/hardforks/merge.ts","./src/common/hardforks/mergeForkIdTransition.ts","./src/common/hardforks/muirGlacier.ts","./src/common/hardforks/petersburg.ts","./src/common/hardforks/shanghai.ts","./src/common/hardforks/sharding.ts","./src/common/hardforks/spuriousDragon.ts","./src/common/hardforks/tangerineWhistle.ts","./src/tx/address.ts","./src/tx/baseTransaction.ts","./src/tx/constants.ts","./src/tx/eip1559Transaction.ts","./src/tx/eip2930Transaction.ts","./src/tx/index.ts","./src/tx/legacyTransaction.ts","./src/tx/transactionFactory.ts","./src/tx/types.ts","./src/tx/utils.ts"],"version":"5.6.2"}
```
Comment (krzysu) [inline packages/web3-eth-accounts/tsconfig.types.tsbuildinfo]: I think all `.tsbuildinfo` files should be added to gitignore

**Review by krzysu — APPROVED** [review]

**Pull request merged**